### PR TITLE
Allow passing the data to the manifest through environment variables

### PR DESCRIPTION
**Pull request by pepibumur**
### Short description 📝
Tyler Neveldine shared with me how they use Tuist to generate white-label apps, and mentioned how cool it'd be if they had a way to pass attributes that could be used from the manifest to change the configuration of the projects. It sounded very feasible so I just went ahead and implemented it. 

### Solution 📦

- I added a new struct to `ProjectDescription`, `Environment`, which leverages `@dynamicMemberLookup` to provide an API for accessing variables.
- Variables are passed following the convention `TUIST_xxxx`. For instance, a variable `TUIST_MY_VARIABLE` is accessed from the manifest calling `Environment.myVariable`.
- I adjusted the manifest loader and the helpers builder to call the Swift compiler passing all the `TUIST_` environment variables so that they can be accessed from the Swift files.

### Implementation 👩‍💻👨‍💻
- [x] Add `Environment` to `ProjectDescription`.
- [x] Change logic `ManifestLoader` and `ProjectDescriptionHelpersBuilder` to pass down all `TUIST_` variables.
- [x] Add an acceptance test that validates that the feature works as intended.
- [x] Adjust the helpers hashing logic to take into account all the `TUIST_` variables.
- [x] Update the documentation.

### CODE REVIEW

File: Sources/ProjectDescription/Environment.swift [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+
+private extension String {
+    func camelCaseToSnakeCase() -> String {
+        let acronymPattern = "([A-Z]+)([A-Z][a-z]|[0-9])"
+        let normalPattern = "([a-z0-9])([A-Z])"
+        return processCamalCaseRegex(pattern: acronymPattern)?
+            .processCamalCaseRegex(pattern: normalPattern)?.lowercased() ?? lowercased()
+    }
+
+    func processCamalCaseRegex(pattern: String) -> String? {
```
Comment (fortmarek) [inline Sources/ProjectDescription/Environment.swift:33]: Just a nitpick: you have a typo here `Camal`

File: Sources/ProjectDescription/Environment.swift [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+            return .boolean(false)
+        } else {
+            return .string(value)
+        }
+    }
+}
+
+private extension String {
+    func camelCaseToSnakeCase() -> String {
+        let acronymPattern = "([A-Z]+)([A-Z][a-z]|[0-9])"
```
Comment (fortmarek) [inline Sources/ProjectDescription/Environment.swift:27]: Regex is probably a fine solution as we want to allow only a limited amount of characters, but if we wanted something more robust, we can copy `JSONEncoder.convertFromSnakeCase` from [here](https://github.com/apple/swift/blob/88b093e9d77d6201935a2c2fb13f27d961836777/stdlib/public/Darwin/Foundation/JSONEncoder.swift#L161)

File: Sources/ProjectDescription/Environment.swift [hunk 3]
```diff
@@ -0,0 +1,38 @@
+import Foundation
+
+@dynamicMemberLookup
```
Comment (fortmarek) [inline Sources/ProjectDescription/Environment.swift:3]: Great use of `@dynamicMemberLookUp` ❤️

File: Sources/ProjectDescription/Environment.swift (hunk 2 above)
Comment (pepibumur) [reply Sources/ProjectDescription/Environment.swift:27]: I like the idea 😛. I've taken the implementation from there.

**Pull request merged**